File: target-dotnet/src/SESS.NexaERP.Domain/Stores/StockAdjustment.cs
namespace SESS.NexaERP.Domain.Stores;

public static class StockAdjustmentStatuses
{
    public const string Draft = "DRAFT";
    public const string Submitted = "SUBMITTED";
    public const string Approved = "APPROVED";
    public const string Posted = "POSTED";
    public const string Rejected = "REJECTED";
}

public static class StockAdjustmentReasonKinds
{
    public const string CountVariance = "COUNT_VARIANCE";
    public const string DamageLoss = "DAMAGE_LOSS";
    public const string Correction = "CORRECTION";
    public static readonly string[] All = [CountVariance, DamageLoss, Correction];
}

/// <summary>
/// A stock adjustment: count variance, damage/loss write-off or correction, recorded by Stores,
/// decided by the roles the approval snapshot names, posted to the physical and FIFO ledgers by
/// the decision that completes the review. Lines and decisions are append-only; a posted
/// adjustment is immutable and corrected only by a reversing adjustment.
/// </summary>
public sealed class StockAdjustment
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid CompanyId { get; set; }
    public string AdjustmentNumber { get; set; } = string.Empty;
    public Guid WarehouseId { get; set; }
    public string ReasonKind { get; set; } = StockAdjustmentReasonKinds.CountVariance;
    public DateOnly EffectiveDate { get; set; }
    public Guid InventoryPeriodId { get; set; }
    public string Status { get; set; } = StockAdjustmentStatuses.Draft;
    public int CurrentRevisionNumber { get; set; } = 1;
    public string Remarks { get; set; } = string.Empty;
    public Guid RecordedByEmployeeId { get; set; }
    public Guid RecordedRoleAssignmentId { get; set; }
    public string CounterEmployeeIdsJson { get; set; } = "[]";
    public string? BackdateReason { get; set; }
    public Guid? BackdateEvidenceId { get; set; }
    public int DaysBackdated { get; set; }
    public string? ApprovalSnapshotJson { get; set; }
    public Guid? ReversesStockAdjustmentId { get; set; }
    public Guid? StockPostingBatchId { get; set; }
    public DateTimeOffset? PostedAt { get; set; }
    public Guid? PostedByEmployeeId { get; set; }
    public string? PostingIdempotencyKey { get; set; }
    public string? PostingRequestFingerprint { get; set; }
    public string IdempotencyKey { get; set; } = string.Empty;
    public string RequestFingerprint { get; set; } = string.Empty;
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public string CreatedBy { get; set; } = "system";
    public DateTimeOffset? UpdatedAt { get; set; }
    public string? UpdatedBy { get; set; }
    public long Version { get; set; }
}

public sealed class StockAdjustmentLine
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid CompanyId { get; set; }
    public Guid StockAdjustmentId { get; set; }
    public int RevisionNumber { get; set; }
    public int LineNumber { get; set; }
    public Guid ItemId { get; set; }
    public Guid WarehouseConditionLocationId { get; set; }
    public string? LotNumber { get; set; }
    public string? SerialNumber { get; set; }
    public decimal QuantityChange { get; set; }
    public decimal? UnitValue { get; set; }
    public decimal AcceptedLineValue { get; set; }
    public string? Remarks { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public string CreatedBy { get; set; } = "system";
}

public sealed class StockAdjustmentDecision
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid CompanyId { get; set; }
    public Guid StockAdjustmentId { get; set; }
    public int RevisionNumber { get; set; }
    public string Decision { get; set; } = "APPROVE";
    public Guid EmployeeId { get; set; }
    public string RoleCode { get; set; } = string.Empty;
    public Guid RoleAssignmentId { get; set; }
    public string RoleAssignmentType { get; set; } = string.Empty;
    public DateTimeOffset DecidedAt { get; set; }
    public string Reason { get; set; } = string.Empty;
    public string IdempotencyKey { get; set; } = string.Empty;
    public string CreatedBy { get; set; } = "system";
}

File: target-dotnet/src/SESS.NexaERP.Web/src/types/stockAdjustment.ts
// A2 stock adjustment wire contract: SESS.NexaERP.Application/Stores/StockAdjustmentContracts.cs.
// PascalCase on the wire (ApiJsonContract). DateOnly fields are yyyy-MM-dd;
// DateTimeOffset fields arrive as ISO instants and are shown in local time.

export const STOCK_ADJUSTMENT_STATES = ['DRAFT', 'SUBMITTED', 'APPROVED', 'POSTED', 'REJECTED'] as const
export type StockAdjustmentStatus = (typeof STOCK_ADJUSTMENT_STATES)[number]

/** StockAdjustmentReasonKinds.All. DAMAGE_LOSS removes stock only (write-off: TD with Accounts concurrence). */
export const STOCK_ADJUSTMENT_REASON_KINDS = ['COUNT_VARIANCE', 'DAMAGE_LOSS', 'CORRECTION'] as const
export type StockAdjustmentReasonKind = (typeof STOCK_ADJUSTMENT_REASON_KINDS)[number]

export const STOCK_ADJUSTMENT_REASON_WORDS: Record<StockAdjustmentReasonKind, string> = {
  COUNT_VARIANCE: 'Count variance — physical count differs from the book',
  DAMAGE_LOSS: 'Damage / loss write-off — removes stock only',
  CORRECTION: 'Correction — an earlier entry was wrong',
}

/** Role codes the approval snapshot can name, in plain words. Display only; authority comes from the server. */
export const STOCK_ADJUSTMENT_ROLE_WORDS: Record<string, string> = {
  STORES_MANAGER: 'Stores Manager',
  TECHNICAL_DIRECTOR: 'Technical Director',
  MANAGING_DIRECTOR: 'Managing Director',
  ACCOUNTS_MANAGER: 'Accounts Manager (concurrence)',
  STORES_EXECUTIVE: 'Stores Executive',
}

export function stockAdjustmentRoleWords(code: string): string {
  return STOCK_ADJUSTMENT_ROLE_WORDS[code] ?? code.replaceAll('_', ' ')
}

/** StockAdjustmentLineInput. QuantityChange is signed: + adds stock, − removes it. */
export interface StockAdjustmentLineInput {
  ItemId: string
  WarehouseConditionLocationId: string
  QuantityChange: number
  /** Required (≥ 0) for an addition; must be null for a removal (valued from FIFO). */
  UnitValue: number | null
  LotNumber: string | null
  /** A serialized line changes exactly one unit. */
  SerialNumber: string | null
  Remarks: string | null
}

export interface CreateStockAdjustmentRequest {
  WarehouseId: string
  ReasonKind: StockAdjustmentReasonKind
  /** DateOnly, yyyy-MM-dd. Inside the chosen open inventory period, not after the server's (UTC) date. */
  EffectiveDate: string
  InventoryPeriodId: string
  Remarks: string
  Lines: StockAdjustmentLineInput[]
  IdempotencyKey: string
  CounterEmployeeIds: string[] | null
  BackdateReason: string | null
  BackdateEvidenceId: string | null
  ReversesStockAdjustmentId: string | null
}

/** PUT: replaces the lines as a new revision and returns the record to DRAFT. */
export interface ReviseStockAdjustmentRequest {
  Version: number
  Remarks: string
  Lines: StockAdjustmentLineInput[]
  Reason: string
  IdempotencyKey: string
  BackdateReason: string | null
  BackdateEvidenceId: string | null
}

export interface StockAdjustmentTransitionRequest {
  Version: number
  Reason: string
  IdempotencyKey: string
}

export interface StockAdjustmentDecisionRequest extends StockAdjustmentTransitionRequest {
  /** Which outstanding required role the approver acts in; null lets the server pick the least privileged one held. */
  RoleCode: string | null
}

export interface StockAdjustmentLineView {
  Id: string
  LineNumber: number
  ItemId: string
  ItemCode: string
  ItemName: string
  WarehouseConditionLocationId: string
  RackBinId: string
  RackBinCode: string
  LotNumber: string | null
  SerialNumber: string | null
  QuantityChange: number
  UnitValue: number | null
  /** Addition: quantity × stated unit value. Removal: FIFO carrying value at recording. */
  AcceptedLineValue: number
  Remarks: string | null
}

export interface StockAdjustmentDecisionView {
  Id: string
  RevisionNumber: number
  Decision: 'APPROVE' | 'REJECT' | string
  EmployeeId: string
  EmployeeCode: string
  EmployeeName: string
  RoleCode: string
  RoleAssignmentId: string
  RoleAssignmentType: string
  DecidedAt: string
  Reason: string
}

export interface StockAdjustmentView {
  Id: string
  AdjustmentNumber: string
  WarehouseId: string
  WarehouseCode: string
  ReasonKind: StockAdjustmentReasonKind | string
  EffectiveDate: string
  InventoryPeriodId: string
  InventoryPeriodCode: string
  Status: StockAdjustmentStatus | string
  CurrentRevisionNumber: number
  Remarks: string
  RecordedByEmployeeId: string
  RecordedByEmployeeCode: string
  CounterEmployeeIds: string[]
  BackdateReason: string | null
  BackdateEvidenceId: string | null
  DaysBackdated: number
  /** Snapshot value once submitted; before that, the sum of the current lines' accepted values. */
  AbsoluteValue: number
  /** From the approval snapshot taken at submission; empty while DRAFT/REJECTED. */
  RequiredRoleCodes: string[]
  /** Required roles not yet approved on the current revision; only while SUBMITTED. */
  OutstandingRoleCodes: string[]
  /** Recorder and counters: they may not approve or reject. */
  ExcludedEmployeeIds: string[]
  ReversesStockAdjustmentId: string | null
  StockPostingBatchId: string | null
  PostedAt: string | null
  Version: number
  Replayed: boolean
  Lines: StockAdjustmentLineView[]
  Decisions: StockAdjustmentDecisionView[]
}

/** StockAdjustmentPage — note Total, not TotalCount. */
export interface StockAdjustmentPage {
  Total: number
  Page: number
  PageSize: number
  Items: StockAdjustmentView[]
}

export interface StockAdjustmentPeriodView {
  Id: string
  Code: string
  Name: string
  StartDate: string
  EndDate: string
  Status: string
}
